File: binary_search/binary_search.rb
# 0(log n) time complexity
# does not take into account sorting--array must be sorted first
def binary_search(list, item)
  low = 0 #first index in array
  high = list.length - 1 #last index of array

  while low <= high
    mid = (low + high)/2 #check middle element
    guess = list[mid] 

    if guess == item
      return mid
    elsif guess > item
      high = mid-1
    else
      low = mid + 1
    end
  end
end

my_list = (1..100).to_a
my_list_2 = (1..10).to_a

puts binary_search(my_list, 20) #will return index value where is found
puts binary_search(my_list_2, 1)

File: binary_search/binary_search.js
// 0(log n)
// should always be sorted first (which can be extra step)
function binarySearch(list, item) {
  let low = 0;
  let high = list.length - 1;
  let mid;
  
  while(low <= high) {
    mid = Math.floor((low + high) / 2);
    let guess = list[mid];
    if (guess === item) {
      return mid;
    } else if (guess > item) {
      high = mid - 1;
    } else {
        low = mid + 1;
    }
  }
  return null;
};

const myList = [1,3,5,7,9,11,13,15];
console.log(binarySearch(myList, 5));
